File: app/src/main/java/com/framgia/movieapp/services/NowPlayingRepository.java
package com.framgia.movieapp.services;

import android.support.annotation.Nullable;
import com.framgia.movieapp.model.NowPlaying;
import rx.Observable;

/**
 * Created by FRAMGIA\pham.duc.nam on 23/06/2017.
 */

public class NowPlayingRepository {
    @Nullable
    private final NowPlayingRemoteSource mRemoteSource;
    private static NowPlayingRepository mInstance;

    public NowPlayingRepository(NowPlayingRemoteSource remoteSource) {
        mRemoteSource = remoteSource;
    }

    public static NowPlayingRepository getInstance(@Nullable NowPlayingRemoteSource remoteSource) {
        if (mInstance == null) {
            mInstance = new NowPlayingRepository(remoteSource);
        }
        return mInstance;
    }

    public Observable<NowPlaying> getNowPlaying(int page) {
        return mRemoteSource.getNowLoading(page);
    }
}

File: app/src/main/java/com/framgia/movieapp/services/NowPlayingRepository.kt
package com.framgia.movieapp.services

import com.framgia.movieapp.model.NowPlaying
import rx.Observable

/**
 * Created by FRAMGIA\pham.duc.nam on 23/06/2017.
 */

class NowPlayingRepository(private val mRemoteSource: NowPlayingRemoteSource?) {

  fun getNowPlaying(page: Int): Observable<NowPlaying> {
    return mRemoteSource!!.getNowLoading(page)
  }

  companion object {
    private var mInstance: NowPlayingRepository? = null

    fun getInstance(remoteSource: NowPlayingRemoteSource?): NowPlayingRepository {
      if (mInstance == null) {
        mInstance = NowPlayingRepository(remoteSource)
      }
      return mInstance as NowPlayingRepository
    }
  }
}
